File: Solutions/day3/py/main.py
def get_most_common(arr: list, pos: int):
    gamma_bit_count = 0
    epsilon_bit_count = 0
    for binary in arr:
        if binary[pos] == "1":
            gamma_bit_count += 1
        else:
            epsilon_bit_count += 1

    if gamma_bit_count >= epsilon_bit_count:
        return "1"
    else:
        return "0"


def flip(bit: str):
    if bit == "1":
        return "0"
    else:
        return "1"


def part_a(arr):
    gamma = ""
    epsilon = ""
    for pos in range(len(arr[0])):
        common = get_most_common(arr, pos)
        gamma += common
        epsilon += flip(common)
    return int(gamma, 2) * int(epsilon, 2)


def get_rating(arr: list, f: bool = False):
    bits_len = len(arr[0])
    result = arr[:]
    for i in range(bits_len):
        common = get_most_common(result, i)
        common = flip(common) if f else common
        result = [_b for _b in result if _b[i] == common]
        if len(result) == 1:
            break
    return result


def part_b(arr):
    oxygen = get_rating(arr[:])
    co2 = get_rating(arr[:], True)
    return int("".join(oxygen[0]), 2) * int("".join(co2[0]), 2)


def main() -> tuple:
    with open("../input.txt") as f:
        arr = [bits[:-1] for bits in f]
    return part_a(arr), part_b(arr)


if __name__ == '__main__':
    import time

    start_time = time.perf_counter()
    a, b = main()
    end_time = time.perf_counter()
    print(f"Result part_a: {a}")
    print(f"Result part_b: {b}")
    print(f"Time: {(end_time - start_time):4f} sec")

File: Solutions/day3/rust/src/main.rs
use std::fs;

fn get_common(arr: &[String], pos: usize) -> &str {
    let mut gamma_count = 0;
    let mut epsilon_count = 0;

    for bits in arr {
        if bits.as_bytes()[pos] == b'1' {
            gamma_count += 1;
        } else {
            epsilon_count += 1;
        }
    }

    if gamma_count >= epsilon_count {
        "1"
    } else {
        "0"
    }
}

fn flip(bit: &str) -> &str {
    if bit == "1" {
        return "0";
    }
    "1"
}

fn str_binary_to_int(bits: String) -> i32 {
    i32::from_str_radix(bits.as_str(), 2).unwrap()
}

fn part_a(v: &Vec<String>) -> i32 {
    let mut gamma = String::new();
    let mut epsilon = String::new();
    let mut common_bit: &str;
    for pos in 0..v.first().unwrap().len() {
        common_bit = get_common(&v, pos);
        gamma += common_bit;
        epsilon += flip(common_bit);
    }
    str_binary_to_int(gamma) * str_binary_to_int(epsilon)
}

fn get_rating(arr: &Vec<String>, f: bool) -> String {
    let bits_len = arr.first().unwrap().len();
    let mut result: Vec<String> = arr.clone();
    let mut common: &str;
    for i in 0..bits_len {
        let cloned_result = &result.clone(); // &result.clone();
        common = get_common(cloned_result, i);
        common = if f { flip(common) } else { common };
        result.retain(|x| x.as_bytes()[i] == common.as_bytes()[0]);
        if result.len() == 1 {
            break;
        }
    }
    result.first().unwrap().to_string()
}

fn part_b(v: &Vec<String>) -> i32 {
    let oxygen = get_rating(v, false);
    let co2 = get_rating(v, true);
    str_binary_to_int(oxygen) * str_binary_to_int(co2)
}

fn main() {
    let timer = std::time::Instant::now();
    let input_vec: Vec<String> = fs::read_to_string("../input.txt")
        .unwrap()
        .lines()
        .map(|x| x.to_string())
        .collect::<Vec<_>>();

    let a = part_a(&input_vec);
    let b = part_b(&input_vec);
    let total_time = timer.elapsed();
    println!("Result part_a: {}", a);
    println!("Result part_b: {}", b);
    println!("Time: {} sec", total_time.as_secs_f32());
}
